tket2-hseries/src/lazify_measure.rs

```diff
@@ -74,158 +85,246 @@ impl LazifyMeasurePass {
+        Ok(Self(SimpleReplacement::new(
+            subgraph,
+            replacement,
+            nu_inp,
+            nu_out,
+        )))
+    }
 
-    // A map from (target ports of edges from the Input node of `replacement`) to (target ports of
-    // edges from nodes not in `removal` to nodes in `removal`).
-    let nu_inp = replacement_hugr
-        .all_linked_inputs(i)
-        .map(|(n, p)| ((n, p), (node, p)))
-        .collect();
+    fn build_futures_gadget(builder: &mut impl Dataflow, wire: Wire, num_uses: usize) -> Vec<Wire> {
+        let future_wires = if num_uses == 0 {
+            builder.add_free(wire, bool_t()).unwrap();
+            vec![]
+        } else {
+            let mut future_wires = vec![wire];
+            for _ in 1..num_uses {
+                let prev_wire = future_wires.last_mut().unwrap();
+                let [wire1, wire2] = builder.add_dup(*prev_wire, bool_t()).unwrap();
+                *prev_wire = wire1;
+                future_wires.push(wire2);
+            }
+            future_wires
+        };
+        debug_assert_eq!(future_wires.len(), num_uses);
 
-    // qubit is linear, there must be exactly one
-    let (target_node, target_port) = hugr
-        .single_linked_input(node, OutgoingPort::from(0))
-        .unwrap();
-    // A map from (target ports of edges from nodes in `removal` to nodes not in `removal`) to
-    // (input ports of the Output node of `replacement`).
-    let mut nu_out: HashMap<_, _> = [((target_node, target_port), IncomingPort::from(0))]
-        .into_iter()
-        .collect();
-    nu_out.extend(
-        hugr.linked_inputs(node, OutgoingPort::from(1))
-            .enumerate()
-            .map(|(i, target)| (target, IncomingPort::from(i + 1))),
-    );
-    assert_eq!(nu_out.len(), 1 + num_uses_of_bool);
-    assert_eq!(nu_out.len(), replacement_hugr.in_value_types(o).count());
+        future_wires
+            .into_iter()
+            .map(|w| builder.add_read(w, bool_t()).unwrap()[0])
+            .collect_vec()
+    }
 
-    let nu_out_set = nu_out.keys().copied().collect();
-    (
-        nu_out_set,
-        SimpleReplacement::new(g, replacement_hugr, nu_inp, nu_out),
-    )
+    fn check_signature(
+        node: Node,
+        qsystem_op: QSystemOp,
+        op_to_replace: &hugr::ops::OpType,
+    ) -> Result<(), LazifyMeasurePassError> {
+        let expected_signature = qsystem_op
+            .to_extension_op()
+            .unwrap()
+            .signature()
+            .into_owned();
+        let actual_signature = op_to_replace
+            .dataflow_signature()
+            .map(std::borrow::Cow::into_owned);
+        if let Some(sig) = &actual_signature {
+            if expected_signature.input() == sig.input()
+                && expected_signature.output().len() == sig.output().len()
+                && itertools::zip_eq(expected_signature.output().iter(), sig.output().iter())
```

Comment:
could you add a comment explaining these conditions? types matching up to bool == future(bool) ?

Reply:
muich